subsys/fs/fcb/Kconfig

```diff
@@ -10,6 +10,7 @@
 config FCB
 	bool "Flash Circular Buffer support"
 	depends on FLASH_MAP
+	depends on FLASH_HAS_PAGE_LAYOUT
```

Comment:
It should not be dependent on that, the `FLASH_PAGE_LAYOUT` should be.
The `FLASH_HAS_PAGE_LAYOUT` should be information for `FLASH_PAGE_LAYOUT` that  it is available for some devices.

Reply:
`FLASH_PAGE_LAYOUT` is already dependent upon `FLASH_HAS_PAGE_LAYOUT`, the problem is that `select` ignores all dependencies...
https://github.com/zephyrproject-rtos/zephyr/blob/d8183d8e2dad009f8d9dc423104a935d82fa348c/drivers/flash/Kconfig#L144-L152
https://docs.zephyrproject.org/latest/build/kconfig/tips.html#select-pitfalls

Reply:
OK, but that is not the solution. Maybe select here is not a proper keyword to use here.
The same can be done where something selects `FLASH` and then you have to add `FLASH_HAS_DRIVER` dependency to that Kconfig; we start doing that, suddenly everything depends on everything it should not even know about.

Reply:
> The same can be done where something selects FLASH and then you have to add FLASH_HAS_DRIVER dependency to that Kconfig; we start doing that, suddenly everything depends on everything it should not even know about.

That's not strictly the case here, these drivers are explicitly using the `FLASH_PAGE_LAYOUT` API, so requiring them to specify `depends on FLASH_HAS_PAGE_LAYOUT` is not some huge burden.

The other solution is to transition the `select FLASH_PAGE_LAYOUT` to `depends on FLASH_PAGE_LAYOUT`, and then transition all `select SUBSYSTEM` usage to `depends on SUBSYSTEM` as well (because `select` ignores dependencies)

Reply:
Maybe I am wrong here, let me think.

Reply:
It looks like there is only one single flash driver that does not `select FLASH_HAS_PAGE_LAYOUT` (`Kconfig.nios2_qspi`).

Isn't all flash hardware organized in pages, so that this API could be implemented for all drivers? Wouldn't it be better to return `-ENOSYS` if an API is not actually implemented and get rid of `FLASH_HAS_PAGE_LAYOUT`? This would be a more complicated solution, though.

I tend to agree that switching to `depends on FLASH_PAGE_LAYOUT` is cleaner, even though it means we need to enable this Kconfig in all application prj.conf files manually.

Reply:
Every flash driver that provided the page layout should advertise the feature with FLASH_HAS_PAGE_LAYOUT. I consider it a bug if it does not.
Not all flash is, because flash API now supports devices based on MRAM and RRAM tech, like Amibq or nrf54l15, which have to emulate layout just for some of subsystems to work (like FCB).
But once we have subsystems ready for working with or without pages, the entire FLAS_PAGE_LAYOUT may not be required in some configurations.

Reply:
> Isn't all flash hardware organized in pages, so that this API could be implemented for all drivers? Wouldn't it be better to return `-ENOSYS` if an API is not actually implemented and get rid of `FLASH_HAS_PAGE_LAYOUT`? This would be a more complicated solution, though.

I do not think that would be more complicated, though the `FLASH_HAS_PAGE_LAYOUT` has only one purpose: to advertise that there is a device that provides the page layout. Currently only the `FLASH_PAGE_LAYOUT` uses this to allow enabling/selecting the api if there is anything to support it.
Just returning `-ENOSYS` from drivers means that every driver will have to either provide page layout or function returning layout, so if you do not need nor use it you still get the dead code that subtracts from your storage area.

`FLASH_HAS_PAGE_LAYOUT` should not be directly selected nor dependent on by any Kconfig options unless that option provided alternative path of execution that is specific for devices that have `FLASH_HAS_PAGE_LAYOUT` or have not.
Things that only work with page layout depend or select `FLASH_PAGE_LAYOUT`, as a required API.